prusti-encoder/src/encoders/mir_impure.rs

```diff
@@ -58,194 +88,57 @@ impl TaskEncoder for MirImpureEnc {
-
-                    current_stmts: None,
-                    current_terminator: None,
-                    encoded_blocks,
-                };
-                visitor.visit_body(&body);
-
-                visitor.encoded_blocks.push(
-                    vcx.mk_cfg_block(
-                        vcx.alloc(vir::CfgBlockLabelData::End),
-                        &[],
-                        vcx.alloc(vir::TerminatorStmtData::Exit)
-                    )
-                );
-                Some(vcx.alloc_slice(&visitor.encoded_blocks))
-            } else {
-                None
-            };
-
-            let spec = deps.require_local::<MirSpecEnc>(
-                (def_id, substs, caller_def_id, false)
-            ).unwrap();
-            let (spec_pres, spec_posts) = (spec.pres, spec.posts);
-
-            let mut pres = Vec::with_capacity(arg_count - 1);
-            let mut args = Vec::with_capacity(arg_count + substs.len());
-            for arg_idx in 0..arg_count {
-                let name_p = local_defs.locals[arg_idx.into()].local.name;
-                args.push(vir::vir_local_decl! { vcx; [name_p] : Ref });
-                if arg_idx != 0 {
-                    pres.push(local_defs.locals[arg_idx.into()].impure_pred);
-                }
-            }
-            args.extend(param_ty_decls.iter());
-            pres.extend(spec_pres);
-
-            let mut posts = Vec::with_capacity(spec_posts.len() + 1);
-            posts.push(local_defs.locals[mir::RETURN_PLACE].impure_pred);
-            posts.extend(spec_posts);
-
-            Ok((
-                MirImpureEncOutput {
-                    method: vcx.mk_method(
-                        method_ref,
-                        vcx.alloc_slice(&args),
-                        &[],
-                        vcx.alloc_slice(&pres),
-                        vcx.alloc_slice(&posts),
-                        blocks,
-                    ),
-                },
-                (),
-            ))
-        })
+    ) -> Result<
+        (
+            Self::OutputFullLocal<'vir>,
+            Self::OutputFullDependency<'vir>,
+        ),
+        (
+            Self::EncodingError,
+            Option<Self::OutputFullDependency<'vir>>,
+        ),
+    > {
+        let monomorphize = Self::monomorphize();
+        let output_ref = if monomorphize {
+            deps.require_ref::<MirMonoImpureEnc>(*task_key).unwrap()
+        } else {
+            deps.require_ref::<MirPolyImpureEnc>(task_key.def_id).unwrap()
+        };
+        deps.emit_output_ref::<Self>(*task_key, output_ref);
+        let output: ImpureFunctionEncOutput<'_> = if monomorphize {
+            deps.require_local::<MirMonoImpureEnc>(*task_key).unwrap()
+        } else {
+            deps.require_local::<MirPolyImpureEnc>(task_key.def_id).unwrap()
+        };
+        Ok((output, ()))
     }
 }
 
-struct EncVisitor<'tcx, 'vir, 'enc>
+pub struct ImpureEncVisitor<'tcx, 'vir, 'enc>
```

Comment:
This struct is here to make the PR diff easier to read, but perhaps later it should be moved to the file in which it is used `encoder_traits/impure_function_enc.rs`

Reply:
Sounds good. I assume this is also why all the fields are `pub`?